app/src/main/java/org/mozilla/fenix/historymetadata/HistoryMetadataMiddleware.kt

```diff
@@ -0,0 +1,110 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.historymetadata
+
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.ContentAction
+import mozilla.components.browser.state.action.HistoryMetadataAction
+import mozilla.components.browser.state.action.MediaSessionAction
+import mozilla.components.browser.state.action.TabListAction
+import mozilla.components.browser.state.selector.findTab
+import mozilla.components.browser.state.selector.selectedTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.TabSessionState
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.lib.state.Store
+
+/**
+ * This [Middleware] reacts to various browsing events and records history metadata as needed.
+ */
+class HistoryMetadataMiddleware(
+    private val historyMetadataService: HistoryMetadataService
+) : Middleware<BrowserState, BrowserAction> {
+
+    @Suppress("ComplexMethod", "NestedBlockDepth")
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        // Pre process actions
+        when (action) {
+            is TabListAction.AddTabAction -> {
```

Comment:
does this need a `private` check? I don't see an action for adding private tabs..

Reply:
same question for `SelectTabAction` and `RemoveTabAction` - i don't see their private equivalent. If this _is_ ok, can you add a comment explaining why?

Reply:
Good catch! Added a bunch new tests and handled in middleware.